Keystore set create

Arsnael opened the pull request.

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/CustomMethod.scala:
```diff
@@ -16,6 +16,7 @@ import reactor.core.scala.publisher.SMono
 object CapabilityIdentifier {
   val LINAGORA_ECHO: CapabilityIdentifier = "com:linagora:params:jmap:echo"
   val LINAGORA_FILTER: CapabilityIdentifier = "com:linagora:params:jmap:filter"
+  val LINAGORA_PGP: CapabilityIdentifier = "com:linagora:params:jmap:pgp"
```
Can we have tests showing this is advertised in the session?

chibenwa replied:
Can we have tests were we call the session endpoint and ensure the capability is well exposed in the session?

I did not see them....

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,76 @@
+                     |			"accountId": "$generateAccountIdAsString",
+                     |			"create": {
+                     |        "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6": {
+                     |          "key": "GPG Armored"
```
Please reuse a real key here

Arsnael replied:
WIP...

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/json/KeystoreSerializer.scala:
```diff
@@ -0,0 +1,58 @@
+          case Right(valueP) => JsSuccess(valueP)
+          case Left(error)   => JsError(error)
+        }
+      })
```
We should have that (refined reads/writes and common utils shared with filter serializer defined in a package.scala class)

Arsnael replied:
Was thinking about it, but it's like in MOB, make it work first, refactor after

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/json/KeystoreSerializer.scala:
```diff
@@ -0,0 +1,58 @@
+  private implicit val keystoreMapCreationResponseWrites: Writes[Map[KeystoreCreationId, KeystoreCreationResponse]] =
+    mapWrites[KeystoreCreationId, KeystoreCreationResponse](_.value, keystoreCreationResponseWrites)
+
+  def serializeFilterGetResponse(response: KeystoreSetResponse): JsValue = Json.toJson(response)
```
Unrelated

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/model/KeystoreSet.scala:
```diff
@@ -0,0 +1,24 @@
+import play.api.libs.json.JsObject
+
+object KeystoreSetRequest {
+  type KeystoreCreationId = String Refined NonEmpty
```
I prefer

```
case class KeystoreCreationId(Id id)
```

See related refactoring in James repo....

Arsnael commented:
Still in progress... Just got some weird NPE at the deserialization of the request... will check monday

chibenwa commented on openpaas-james/integration-tests/jmap/memory-jmap-integration-tests/src/test/java/com/linagora/openpaas/james/MemoryLinagoraKeystoreSetMethodTest.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public String generateAccountIdAsString() {
+        return "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6";
+    }
```
Why do we need this?

Arsnael replied:
We need bob's account id? Or we gonna get an error from the request if the account id is wrong? I just took the pattern of other test suites TBH

chibenwa replied:
The accoutn id is a constant, can be hardcoded and do not depend of the underlying storage technology...

Arsnael replied:
Probably can do some refactorings on other test classes then

chibenwa commented on openpaas-james/mailbox/encrypted/api/src/main/scala/com/linagora/openpaas/encrypted/KeystoreManagerContract.scala:
```diff
@@ -8,6 +8,7 @@ import reactor.core.scala.publisher.{SFlux, SMono}
 
 import scala.jdk.CollectionConverters._
 
+// Remark: why is this not in test folder? Need to move =>
```
:+1: and we thus need to fix pom.xml as well...

chibenwa replied:
IMO it deserve a separate PR

Arsnael replied:
#113

Arsnael commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -63,10 +111,10 @@ trait LinagoraKeystoreSetMethodContract {
+         |            "id": "$PGP_KEY_ID"
          |          }
          |				}
          |			}, "c1"]
```
TODO: more tests

Arsnael commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -63,10 +111,10 @@ trait LinagoraKeystoreSetMethodContract {
-         |          "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6": {
-         |            "id": "1"
+         |          "K87": {
+         |            "id": "$PGP_KEY_ID"
```
I put a hardcoded key id now. I saw it was generated same keyid for same key in memory. I would not be surprised that the algorithm might be different for Cassandra key id? Not sure of the approach here

chibenwa replied:
> Cassandra key id?

It will use the key fingerprint as well.

Arsnael commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -33,10 +81,10 @@ trait LinagoraKeystoreSetMethodContract {
-                     |        "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6": {
-                     |          "key": "GPG Armored"
+                     |        "K87": {
+                     |          "key": "$PGP_KEY_BASE_64"
```
In base64 because well you can really pass an array of bytes in an other way as json it seems...

chibenwa replied:
Why do we need this conversion:

```
armored key (text) -> bytes -> base64 (text) ?
```

Can't we have the raw armored key text passed here?

Arsnael replied:
Then let me know how to deal with that with play json and deserialize it cause that caused some json parse exception to me and I could not find a workaround

chibenwa replied:
> Then let me know how to deal with that with play json and deserialize it cause that caused some json parse exception to me and I could not find a workaround

Technical constraint is a bad reason to design a bad API, Rene.

A String is a string be it base64 or armored. I suspect you rather have problems with scala interpolation and line breaks.

Arsnael replied:
I guess it's probably better towards what we want but saying it would be a bad API... I don't see how. 

Anyway, I spent a good amount of time on that already and can't make it work that way, json serialization is driving me nuts. So support on this would be appreciated.

Arsnael commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetCreatePerformer.scala:
```diff
@@ -64,7 +65,7 @@ class KeystoreSetCreatePerformer @Inject()(serializer: KeystoreSerializer,
                         clientId: KeystoreCreationId,
                         keystoreCreationRequest: KeystoreCreationRequest): SMono[KeystoreCreationResult] =
-      SMono.fromPublisher(keystore.save(mailboxSession.getUser, keystoreCreationRequest.key))
+      SMono.fromPublisher(keystore.save(mailboxSession.getUser, Base64.decode(keystoreCreationRequest.keyBase64)))
```
That was done too fast, I should refactor this base64 decode part I think

chibenwa replied:
You should rather get rid of this base64 stuff indeed.

Arsnael replied:
Same as above then, if you manage to pass the key as text in your json let me know how

Arsnael reviewed (commented):
At least the test works now... I need to refactor my code and add more tests

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -1,19 +1,67 @@
 import org.junit.jupiter.api.{BeforeEach, Test}
 
+object LinagoraKeystoreSetMethodContract {
+  private val PGP_KEY: Array[Byte] = s"""-----BEGIN PGP PUBLIC KEY BLOCK-----
```
Use pgp-utils test jar resources?

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,124 @@
+                           |Gp6Bcl9C5IVVPj98cNMDAo5r3k+5vXmKuXtX4aXkO4xE4TKf1jK+4uVI+kj3qP5V
+                           |x6e6cS5SeRbzaDlNKIjjEw==
+                           |=Q4mr
+                           |-----END PGP PUBLIC KEY BLOCK-----""".stripMargin.getBytes
```
Get bytes should ALWAYS take a charset argument...

Arsnael commented:
I think most of the things are here... Rest the question of armored string vs base64... I will try to dig more on it now

Arsnael commented:
Regarding passing the armored key in plain... I am running into something like in this post: https://stackoverflow.com/questions/31537153/jsonparseexception-illegal-unquoted-character-ctrl-char-code-10

```
{
    "type": "urn:ietf:params:jmap:error:notJSON",
    "status": 400,
    "detail": "The content type of the request was not application/json or the request did not parse as I-JSON: Illegal unquoted character ((CTRL-CHAR, code 10)): has to be escaped using backslash to be included in string value\n at [Source: (reactor.netty.ByteBufMono$ReleasingInputStream); line: 8, column: 56]"
}
```

It seems to be the line returns `\n`... I'm honestly not sure we should expect the user modifying his pgp key to escape illegal json character... A conversion in base64 is much simpler? Even Google seems to use it for uploading pgp keys, so I don't find that particularly crazy: https://developers.google.com/standard-payments/reference/pgp-encryption-details

Or if you have a way to conf play json for this, I'm listening :)

Arsnael commented:
```
[ERROR] connectJMAPServerShouldRespondBadRequest{GuiceJamesServer}  Time elapsed: 1.987 s  <<< ERROR!
java.net.SocketException: Connection reset
```
That's an odd one... I have the same locally so I don't think the CI is crazy... Something wrong in the bindings maybe?

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,514 @@
+  }
+
+  @Test
+  @Disabled("Not sure to keep that, and throws actually a server error instead of a method error")
```
Let's rather get rid of base64 in the first place!

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/method/KeystoreSetMethod.scala:
```diff
@@ -0,0 +1,80 @@
+  private def capability(): Capability = KeystoreCapability
+}
+
+class KeystoreSetMethodModule extends AbstractModule {
```
You do not bind KeystoreCapability in the module :-p

Arsnael replied:
The submarine effect I guess...

Arsnael replied:
No I think I do? Below line:
```
install(new KeystoreCapabilitiesModule())
```
???

chibenwa replied:
Ok but I still miss the test.

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/model/KeystoreSet.scala:
```diff
@@ -0,0 +1,53 @@
+import play.api.libs.json.JsObject
+
+object KeystoreSetRequest {
+  type KeystoreCreationId = Id
```
Please create a dedicated case class:

```
case class KeystoreCreationId(Id id)
```

chibenwa commented on openpaas-james/jmap/extensions/src/main/scala/com/linagora/openpaas/james/jmap/model/KeystoreSet.scala:
```diff
@@ -0,0 +1,53 @@
+    }))
+}
+
+case class KeystoreCreationRequest(key: String)
```
String? Strong typing?

chibenwa commented:
```
[ERROR]   MemoryServerTest.connectJMAPServerShouldRespondBadRequest » Socket Connection ...
[INFO] 
[ERROR] Tests run: 6, Failures: 0, Errors: 1, Skipped: 2
```

Arsnael commented:
Could not find what is wrong, but is that test making sense at all btw for us here? I rechecked :

```
@Test
    default void connectJMAPServerShouldRespondBadRequest(GuiceJamesServer server) {
        RestAssured.requestSpecification = requestSpec(server);
        given()
            .body("{\"badAttributeName\": \"value\"}")
        .when()
            .post("/authentication")
        .then()
            .statusCode(400);
    }
```

We are checking an endpoint on jmap-draft, not even jmap-rfc... which makes no sense for tmail no?

Arsnael commented:
Ok so if I start a docker memory james server, and try to connect to jmap, I get this:
```
10:01:47.658 [WARN ] i.n.c.ChannelInitializer - Failed to initialize a channel. Closing: [id: 0xa3f4480d, L:/172.17.0.3:80 - R:/172.17.0.1:42400]
java.lang.ClassNotFoundException: reactor.core.publisher.Sinks$Empty
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 30 common frames omitted
Wrapped by: java.lang.NoClassDefFoundError: Lreactor/core/publisher/Sinks$Empty;
        at java.base/java.lang.Class.getDeclaredFields0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredFields(Class.java:3061)
        at java.base/java.lang.Class.getDeclaredField(Class.java:2409)
        at java.base/java.util.concurrent.atomic.AtomicReferenceFieldUpdater$AtomicReferenceFieldUpdaterImpl$1.run(AtomicReferenceFieldUpdater.java:335)
        at java.base/java.util.concurrent.atomic.AtomicReferenceFieldUpdater$AtomicReferenceFieldUpdaterImpl$1.run(AtomicReferenceFieldUpdater.java:333)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.util.concurrent.atomic.AtomicReferenceFieldUpdater$AtomicReferenceFieldUpdaterImpl.<init>(AtomicReferenceFieldUpdater.java:332)
        at java.base/java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater(AtomicReferenceFieldUpdater.java:116)
        at reactor.netty.channel.ChannelOperations.<clinit>(ChannelOperations.java:554)
        at reactor.netty.channel.ChannelOperations$OnSetup.empty(ChannelOperations.java:528)
        at reactor.netty.transport.TransportConfig.channelOperationsProvider(TransportConfig.java:116)
        at reactor.netty.transport.TransportConfig$TransportChannelInitializer.initChannel(TransportConfig.java:384)
        at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)
        at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)
        at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:938)
        at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:609)
        at io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:46)
        at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1463)
        at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1115)
        at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:650)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:502)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:417)
        at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:474)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
```
From the encrypter? or something else? I will dig more...

chibenwa reviewed (changes requested):
Base 64 -> nope

chibenwa commented:
> From the encrypter? or something else? I will dig more...

Could be a dependency clash, could be a missing dependency.

Verify that the reacttor versions are alligned, run a dependency dump to see if there is some clash, run a git bisect(?) etc...

chibenwa commented:
BTW it is weird that it works in integration tests and not in the memory server...

You may want to hunt the missing/wrong dep that way.

Arsnael commented:
Very weird and driving me crazy... I hate that maven deps mess TBH. 

It would be hard to do a git bisect as without the guice wirings it's meaningless (it comes from the keystore bindings for sure, the tests on master are green)

And I know for the base64 (I didn't mark those remarks as resolved for a reason) but I didn't have the time for it yet.......

Arsnael commented on openpaas-james/pom.xml:
```diff
@@ -706,6 +706,13 @@
+                <version>2020.0.3</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
```
That solves the deployment, but still not the MemoryServerTest where I still get a connection reset (different issue).

What I don't understand is that we use exactly the same setup for integration tests and there is no issue on those...

Arsnael replied:
Ok it solves it it seems, in command line it's fine... My IDE seems quite buggy since it updated to the latest version... it affects my dev honestly. 

Will deal with the armored key instead of the base64 in the json next week...

Arsnael commented:
Cleaned up the history and squashed, except for the last commit, containing the last change asked about armored key in json. It works now :) (needed a weekend to clear out the head I guess)

chibenwa commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,489 @@
+    .when()
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .headers(getHeadersWith(BOB_BASIC_AUTH_HEADER))
+      .get("/session").prettyPeek()
```
Debug spotted

chibenwa commented on openpaas-james/mailbox/encrypted/api/src/test/scala/com/linagora/openpaas/encrypted/KeystoreManagerContract.scala:
```diff
@@ -8,6 +8,7 @@ import reactor.core.scala.publisher.{SFlux, SMono}
 
 import scala.jdk.CollectionConverters._
 
+// Remark: why is this not in test folder? Need to move =>
```
Rebase on master?

(and remove this comment?)

Arsnael replied:
Pretty sure I did remove it in a fixup, probably came back when I rebased... Thanks for spotting this

Arsnael replied:
and the PR was already rebased on master :)

Arsnael commented:
Squashed and rebased

chibenwa commented:
One more review and good to go?

Cc @LanKhuat ...

Arsnael commented:
Rebased to solve small conflict

LanKhuat commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,489 @@
+  def keystoreSetCreateShouldSucceedWithValidKey(): Unit = {
+    val request = s"""{
+                     |	"using": ["urn:ietf:params:jmap:core", "com:linagora:params:jmap:pgp"],
+                     |	"methodCalls": [
+                     |		["Keystore/set", {
+                     |			"accountId": "$ACCOUNT_ID",
+                     |			"create": {
+                     |        "K87": {
+                     |          "key": "$PGP_KEY_ARMORED"
+                     |        }
+                     |      }
+                     |		}, "c1"]
+                     |	]
+                     |}""".stripMargin
```
weird indent here?

LanKhuat replied:
and many other places...

Arsnael replied:
weird... Fine locally... with intellij and visual code... not sure if it is github or my IDEs being crazy here oO

LanKhuat commented on openpaas-james/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/openpaas/james/common/LinagoraKeystoreSetMethodContract.scala:
```diff
@@ -0,0 +1,489 @@
+      .inPath("capabilities.com:linagora:params:jmap:pgp")
+      .isEqualTo("{}")
+  }
+}
```
Shouldn't we assert the content of the saved key(s) (by using probe, repository...) and replace the assertion with Keystore/get when it is available later?
